File: scripts/extract_text.py
"""
Extract plain text from a resume/cover-letter file for read_document.

Usage:  python extract_text.py <file.pdf|file.docx>
Prints a JSON object to stdout: {"text": ..., "words": N, "chars": N, "pages": N?}
Exits non-zero with a stderr message on unsupported type or a missing library
(the caller turns that into a friendly error).
"""
import json
import os
import sys


def extract_pdf(path):
    from pypdf import PdfReader

    reader = PdfReader(path)
    pages = len(reader.pages)
    text = "\n".join((p.extract_text() or "") for p in reader.pages)
    return text, {"pages": pages}


def extract_docx(path):
    from docx import Document

    doc = Document(path)
    parts = [p.text for p in doc.paragraphs]
    for table in doc.tables:
        for row in table.rows:
            for cell in row.cells:
                if cell.text:
                    parts.append(cell.text)
    return "\n".join(parts), {}


def main():
    if len(sys.argv) < 2:
        sys.stderr.write("usage: extract_text.py <file>\n")
        return 2
    path = sys.argv[1]
    ext = os.path.splitext(path)[1].lower()
    try:
        if ext == ".pdf":
            text, meta = extract_pdf(path)
        elif ext == ".docx":
            text, meta = extract_docx(path)
        else:
            sys.stderr.write(f"unsupported file type: {ext or '(none)'}\n")
            return 2
    except ImportError as e:
        sys.stderr.write(f"missing extraction library: {e}\n")
        return 3
    except Exception as e:  # noqa: BLE001
        sys.stderr.write(f"could not extract text: {e}\n")
        return 4

    # Collapse runs of blank lines so the output stays readable.
    lines = [ln.rstrip() for ln in text.splitlines()]
    cleaned, blank = [], False
    for ln in lines:
        if ln.strip() == "":
            if not blank:
                cleaned.append("")
            blank = True
        else:
            cleaned.append(ln)
            blank = False
    text = "\n".join(cleaned).strip()

    out = {"text": text, "words": len(text.split()), "chars": len(text)}
    out.update(meta)
    sys.stdout.write(json.dumps(out))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/extractText.ts
/**
 * Extract plain text from a resume/cover-letter file, for read_document.
 *
 * Ported from the former scripts/extract_text.py (pypdf + python-docx) to
 * pure JS (pdf-parse + mammoth) so this server has no Python dependency.
 */

import * as fs from "node:fs";
// pdf-parse@1.x is CJS (`module.exports = function pdf(buffer, opts) {...}`);
// Node's ESM/CJS interop gives us that function as the default import.
import pdfParse from "pdf-parse";
import * as mammoth from "mammoth";

export interface ExtractedText {
  text: string;
  words: number;
  chars: number;
  pages?: number;
}

/** Collapse runs of blank lines so the output stays readable — same cleanup
 *  extract_text.py did. */
function collapseBlankLines(text: string): string {
  const lines = text.split(/\r\n|\r|\n/).map((l) => l.replace(/\s+$/, ""));
  const cleaned: string[] = [];
  let blank = false;
  for (const ln of lines) {
    if (ln.trim() === "") {
      if (!blank) cleaned.push("");
      blank = true;
    } else {
      cleaned.push(ln);
      blank = false;
    }
  }
  return cleaned.join("\n").trim();
}

async function extractPdf(filePath: string): Promise<{ text: string; pages: number }> {
  const buf = fs.readFileSync(filePath);
  const data = await pdfParse(buf);
  return { text: data.text ?? "", pages: data.numpages ?? 0 };
}

async function extractDocx(filePath: string): Promise<{ text: string }> {
  const result = await mammoth.extractRawText({ path: filePath });
  return { text: result.value ?? "" };
}

/**
 * Extract plain text + metadata from a .pdf or .docx file. Throws a plain
 * Error with a clear message on an unsupported type or extraction failure;
 * the caller (read_document) wraps that into a UserFacingError.
 */
export async function extractText(filePath: string): Promise<ExtractedText> {
  const dot = filePath.lastIndexOf(".");
  const ext = dot >= 0 ? filePath.slice(dot).toLowerCase() : "";

  let raw: string;
  let pages: number | undefined;
  if (ext === ".pdf") {
    const r = await extractPdf(filePath);
    raw = r.text;
    pages = r.pages;
  } else if (ext === ".docx") {
    raw = (await extractDocx(filePath)).text;
  } else {
    throw new Error(`unsupported file type: ${ext || "(none)"}`);
  }

  const text = collapseBlankLines(raw);
  return {
    text,
    words: text ? text.split(/\s+/).filter(Boolean).length : 0,
    chars: text.length,
    ...(pages !== undefined ? { pages } : {}),
  };
}
